### DISCUSSION TITLE
CI/CD Production Deployment strategies for self-hosted Hasura?

### DISCUSSION THREAD

**Discussion by seanonthenet** [hasura/graphql-engine#8447] [General]
Hey guys. 👋 
Can any of you with successful self-hosted (not cloud) production workflows please let me know how you manage them? I can’t find much info on this and have been struggling for a few days with render.com because of database name clashes on their system.

• What CI/CD methods or pipelines do you use? 
• Do you use cli-migrations-v3 docker image or something else? 
• Do you use a CI system like GitHub Actions / CircleCI or do you use whatever your host eg: render.com, DigitalOcean, etc provides to deploy Hasura when you push to your repo?
• How do you manage your metadata and migrations application on changes?
• Is your first push to production different from subsequent pushes regarding metadata and migrations or the same?

There isn’t enough official info about this I think. Any info is appreciated. 🙏

**Comment by poita66** [comment]
I've successfully used cli-migrations-v3 base image with a few CI systems, including Bitbucket Pipelines and GoCD. From a CI perspective, it's just building a Docker image.
You'll need a Dockerfile like:

```Dockerfile
FROM hasura/graphql-engine:v2.6.1.cli-migrations-v3

COPY metadata /hasura-metadata
COPY migrations /hasura-migrations
```

Then you deploy this custom image and the metadata & migrations will be applied when the container starts.

The migrations & metadata get updated when you're using the Hasura Console web UI, and can be checked into git and merged as they're just YAML files.

The first push to production is the same as any other push (provided you have migrations that describe the database from nothing if you're using migrations).

Hope that helps!
